### DISCUSSION TITLE
Using Shear function in perl with alpha channel

### DISCUSSION THREAD

**Discussion by MikeTough** [ImageMagick/ImageMagick#5873] [Help] [answered]
Hi,

I'm trying to use the Shear function in perl on a RGBA file where the alpha channel is 16bit.

The resultant sheared file appears to apply the alpha channel (in conjunction with the fill colour) to the output rather than simply applying the shear to all channels. Is this expected behaviour?

I've tried setting the alpha channel properties prior to applying Shear, but all options in the perl documentation (On, Off, Opaque, Transparent, Copy, Extract, Set) either results in the destruction of the original alpha channel or the alpha channel being applied to the output.

Obviously I could separate out the alpha channel, apply the Shear separately to each component and re-assemble, but is there a smarter way of doing this?

**Comment by snibgo** [comment, marked as answer]
Set the background to "none" before doing the shear.

If that doesn't fix the problem, please state the version of IM, and show a sample input file.

**Reply by MikeTough** [reply]
Thanks Snibgo for your fast and accurate response. This resolves my issue.

Thankfully the sheared space is filled in with black, which is what I'm after. I'm not sure how I would resolve this if I required a different fill colour?

It does seem odd to me that a geometrical function should apply the alpha channel. As far as I can see, none of the other geometrical transforms do this (although I have by no means tried them all).

**Comment by snibgo** [comment]
> Thankfully the sheared space is filled in with black, which is what I'm after.

The sheared space will be filled with "none", which is *transparent* black. Perhaps you are saving to a format that doesn't record transparency, such as JPG. If you want some opaque colour such as "red", then "-background Red -layers Flatten" before writing the output.

> It does seem odd to me that a geometrical function should apply the alpha channel. As far as I can see, none of the other geometrical transforms do this (although I have by no means tried them all).

What do you mean by "apply the alpha channel"? Many geometric transformations calculate each output pixel from more than one input pixel. In those calculations, input pixel colour values are multiplied by their alpha, then the output colour values are calculated as a weighted average, and divided by the weighted average alpha. Hence any input pixels that are fully transparent (with any colour) will not influence output colour values.
